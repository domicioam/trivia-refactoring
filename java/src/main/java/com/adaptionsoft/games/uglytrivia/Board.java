package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;

public abstract class Board {

	public Board() {
		super();
	}

	protected abstract void updatePlayerPosition(int roll, Player player);

	protected abstract int getNextPosition(ArrayList<Player> players, Player currentPlayer);

	public abstract QuestionCategory currentCategory(int playerPosition);
	
	public static Board createBoardBasedOnSize(int numberOfPositions, int numberOfCategories) {
		if(numberOfPositions == 12 && numberOfCategories == 4) {
			return new TwelvePositionsBoard();
		}
		
		throw new UnsupportedOperationException("Game doesn't support a board with the size specified.");
	}

}