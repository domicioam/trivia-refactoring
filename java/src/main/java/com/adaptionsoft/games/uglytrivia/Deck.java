package com.adaptionsoft.games.uglytrivia;

import java.util.LinkedList;

public class Deck {
	public Deck() {
		initializeQuestions();
	}
	
	LinkedList<Question> popQuestions = new LinkedList<Question>();
	LinkedList<Question> scienceQuestions = new LinkedList<Question>();
	LinkedList<Question> sportsQuestions = new LinkedList<Question>();
	LinkedList<Question> rockQuestions = new LinkedList<Question>();

	public void initializeQuestions() {
		for (int i = 0; i < 50; i++) {
			popQuestions.addLast(new Question("Pop Question " + i));
			scienceQuestions.addLast(new Question("Science Question " + i));
			sportsQuestions.addLast(new Question("Sports Question " + i));
			rockQuestions.addLast(new Question("Rock Question " + i));
		}
	}
	
	public void askQuestion(int playerPosition, QuestionCategory currentCategory) {
		System.out.println("The category is " + currentCategory);
		
		switch (currentCategory) {
		case Pop:
			System.out.println(popQuestions.removeFirst());
			break;
		case Science:
			System.out.println(scienceQuestions.removeFirst());
			break;
		case Sports:
			System.out.println(sportsQuestions.removeFirst());
			break;
		case Rock:
			System.out.println(rockQuestions.removeFirst());
			break;
		}
	}
}
