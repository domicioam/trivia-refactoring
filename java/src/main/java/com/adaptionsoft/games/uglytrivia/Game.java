package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;

public class Game {
	ArrayList<Player> players = new ArrayList<Player>();
	boolean isGettingOutOfPenaltyBox;
	private Deck deck;
	private PenaltyBox penaltyBox;
	Player currentPlayer;

	public Game() {
		penaltyBox = new PenaltyBox();
		deck = new Deck();
	}

	public boolean isPlayable() {
		return (howManyPlayers() >= 2);
	}

	public boolean addPlayer(String playerName) {
		Player player = new Player(playerName, 0);
		players.add(player);
		player.setCoins(0);
		
		if(currentPlayer == null) {
			currentPlayer = player;
		}

		System.out.println(playerName + " was added");
		System.out.println("They are player number " + players.size());
		return true;
	}

	public int howManyPlayers() {
		return players.size();
	}

	public void roll(int roll) {
		System.out.println(currentPlayer.getName() + " is the current player");
		System.out.println("They have rolled a " + roll);

		if (penaltyBox.contains(currentPlayer)) {
			isGettingOutOfPenaltyBox = roll % 2 != 0;
			if (isGettingOutOfPenaltyBox) {
				System.out.println(currentPlayer.getName() + " is getting out of the penalty box");
				updatePlayerPositionAndAskQuestion(roll, currentPlayer);
			} else {
				System.out.println(currentPlayer.getName() + " is not getting out of the penalty box");
			}
		} else {
			updatePlayerPositionAndAskQuestion(roll, currentPlayer);
		}
	}

	public boolean wasCorrectlyAnswered() {
		boolean playerDidWin;
		if (penaltyBox.contains(currentPlayer)) {
			if (isGettingOutOfPenaltyBox) {
				increasePlayerCoins(currentPlayer);
				playerDidWin = didPlayerWin();
			} else {
				playerDidWin = true;
			}
		} else {
			increasePlayerCoins(currentPlayer);
			playerDidWin = didPlayerWin();
		}
		
		updateCurrentPlayer();
		return playerDidWin;
	}
	
	public boolean wrongAnswer() {
		System.out.println("Question was incorrectly answered");
		System.out.println(currentPlayer.getName() + " was sent to the penalty box");
		penaltyBox.insertPlayer(currentPlayer);

		updateCurrentPlayer();
		return true;
	}
	
	private void updatePlayerPositionAndAskQuestion(int roll, Player player) {
		updatePlayerPosition(roll, player);
		deck.askQuestion(player.getPosition());
	}

	private void updatePlayerPosition(int roll, Player player) {
		player.setPosition(player.getPosition() + roll);
		int maxPosition = 11;
		if (player.getPosition() > maxPosition) {
			resetPlayerPosition(player);
		}
		System.out.println(player.getName() + "'s new location is " + player.getPosition());
	}

	private void resetPlayerPosition(Player player) {
		player.setPosition(player.getPosition() - 12);
	}

	private void updateCurrentPlayer() {
		int currentPosition = players.indexOf(currentPlayer);
		int nextPosition = currentPosition + 1;
		if (nextPosition > players.size() - 1) {
			nextPosition = 0;
		}
		currentPlayer = players.get(nextPosition);
	}

	private void increasePlayerCoins(Player player) {
		System.out.println("Answer was correct!!!!");
		player.setCoins(player.getCoins() + 1);
		System.out.println(currentPlayer.getName() + " now has " + currentPlayer.getCoins() + " Gold Coins.");
	}

	private boolean didPlayerWin() {
		return !(currentPlayer.getCoins() == 6);
	}
}
