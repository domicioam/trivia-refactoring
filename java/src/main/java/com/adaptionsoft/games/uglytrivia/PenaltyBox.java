package com.adaptionsoft.games.uglytrivia;

public class PenaltyBox {
	boolean[] inPenaltyBox = new boolean[6];
	
	public void insertPlayer(int player) {
		inPenaltyBox[player] = true;
	}
	
	public void removePlayer(int player) {
		inPenaltyBox[player] = false;
	}
	
	public boolean contains(int player) {
		return inPenaltyBox[player];
	}
}
