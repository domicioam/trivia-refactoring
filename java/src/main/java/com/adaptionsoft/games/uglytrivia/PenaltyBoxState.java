package com.adaptionsoft.games.uglytrivia;

public abstract class PenaltyBoxState {
	protected Game game;
	
	public PenaltyBoxState(Game game) {
		this.game = game;
	}
	
	abstract void roll(Player player, int roll);
	abstract boolean wasCorrectlyAnswered(Player player, PenaltyBox penaltyBox);
}
