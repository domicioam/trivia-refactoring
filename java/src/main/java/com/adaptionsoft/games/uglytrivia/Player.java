package com.adaptionsoft.games.uglytrivia;

public class Player {
	public Player(String name, int position) {
		this.name = name;
		this.position = position;
	}
	
	private String name;
	private int position;
	private int coins;

	public String getName() {
		return name;
	}

	public int getPosition() {
		return position;
	}

	public void setPosition(int position) {
		this.position = position;
	}

	public int getCoins() {
		return coins;
	}

	public void setCoins(int coins) {
		this.coins = coins;
	}
}
