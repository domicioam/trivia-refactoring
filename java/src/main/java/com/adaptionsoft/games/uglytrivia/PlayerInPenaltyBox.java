package com.adaptionsoft.games.uglytrivia;

public class PlayerInPenaltyBox extends PenaltyBoxState {

	public PlayerInPenaltyBox(Game game) {
		super(game);
	}

	boolean isGettingOutOfPenaltyBox;
	
	@Override
	public void roll(Player player, int roll) {
		isGettingOutOfPenaltyBox = roll % 2 != 0;
		if (isGettingOutOfPenaltyBox) {
			System.out.println(player.getName() + " is getting out of the penalty box");
			game.updatePlayerPositionAndAskQuestion(roll, player);
		} else {
			System.out.println(player.getName() + " is not getting out of the penalty box");
		}
	}
	
	@Override
	public boolean wasCorrectlyAnswered(Player player, PenaltyBox penaltyBox) {
		boolean playerDidWin = false;
		if (penaltyBox.contains(player)) {
			if (isGettingOutOfPenaltyBox) {
				game.increasePlayerCoins(player);
				playerDidWin = game.didPlayerWin();
			} else {
				playerDidWin = true;
			}
		} 
		
		return playerDidWin;
	}
}
