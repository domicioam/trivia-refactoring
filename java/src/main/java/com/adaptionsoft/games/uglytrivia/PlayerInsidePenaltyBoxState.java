package com.adaptionsoft.games.uglytrivia;

public class PlayerInsidePenaltyBoxState extends PenaltyBoxState {

	public PlayerInsidePenaltyBoxState(Game game) {
		super(game);
	}

	boolean isGettingOutOfPenaltyBox;
	
	@Override
	public void roll(Player player, int roll) {
		isGettingOutOfPenaltyBox = roll % 2 != 0;
		if (isGettingOutOfPenaltyBox) {
			System.out.println(player.getName() + " is getting out of the penalty box");
			game.updatePlayerPositionAndAskQuestion(roll, player);
		} else {
			System.out.println(player.getName() + " is not getting out of the penalty box");
		}
	}
	
	@Override
	public void wasCorrectlyAnswered(Player player, PenaltyBox penaltyBox) {
		if (penaltyBox.contains(player)) {
			if (isGettingOutOfPenaltyBox) {
				game.increasePlayerCoins(player);
				penaltyBox.removePlayer(player);
			} else {
			}
		} 
	}
}
