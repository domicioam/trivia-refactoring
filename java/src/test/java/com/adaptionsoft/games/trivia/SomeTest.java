package com.adaptionsoft.games.trivia;

import static org.junit.Assert.*;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.Test;

import com.adaptionsoft.games.uglytrivia.Game;

public class SomeTest {

	@Test
	public void shouldBePlayableWhenGameHasTwoPlayers() {
		ByteArrayOutputStream byteOutputStream = setCustomOutStream();
		
		Game game = new Game();
		addTwoPlayers(game);
		assertTrue(game.isPlayable());
		assertEquals(2, game.howManyPlayers());
		
		String output = getOutput(byteOutputStream);
		Approvals.verify(output);
	}
	
	@Test
	public void shouldNotBePlayableWhenGameHasOnePlayer() {
		ByteArrayOutputStream byteOutputStream = setCustomOutStream();
		
		Game game = new Game();
		game.add("Player 1");
		assertFalse(game.isPlayable());
		assertEquals(1, game.howManyPlayers());
		
		String output = getOutput(byteOutputStream);
		Approvals.verify(output);
	}
	
	@Test
	public void testTwoTurnsWithOneWrongAnswer() {
		ByteArrayOutputStream byteOutputStream = setCustomOutStream();
		
		Game game = new Game();
		addTwoPlayers(game);
		// first turn
		game.roll(1);
		game.wasCorrectlyAnswered();
		game.roll(1);
		game.wrongAnswer();
		// second turn
		game.roll(1);
		game.wasCorrectlyAnswered();
		game.roll(1);
		game.wasCorrectlyAnswered();
		
		String output = getOutput(byteOutputStream);
		
		Approvals.verify(output);
	}
	
	@Test
	public void testTwoTurnsWithOneWrongAnswerButRollEvenNumber() {
		ByteArrayOutputStream byteOutputStream = setCustomOutStream();
		
		Game game = new Game();
		addTwoPlayers(game);
		// first turn
		game.roll(1);
		game.wasCorrectlyAnswered();
		game.roll(1);
		game.wrongAnswer();
		// second turn
		game.roll(1);
		game.wasCorrectlyAnswered();
		game.roll(2);
		game.wasCorrectlyAnswered();
		
		String output = getOutput(byteOutputStream);
		
		Approvals.verify(output);
	}
	
	@Test
	public void testTwoTurnsWithOneWrongAnswerAndRollEqualEleven() {
		ByteArrayOutputStream byteOutputStream = setCustomOutStream();
		
		Game game = new Game();
		addTwoPlayers(game);
		// first turn
		game.roll(11);
		game.wasCorrectlyAnswered();
		game.roll(1);
		game.wrongAnswer();
		// second turn
		game.roll(1);
		game.wasCorrectlyAnswered();
		game.roll(11);
		game.wasCorrectlyAnswered();
		
		String output = getOutput(byteOutputStream);
		
		Approvals.verify(output);
	}
	
	@Test
	public void testTwoTurnsWithOneWrongAnswerForFirstPlayerAndRollEqualEleven() {
		ByteArrayOutputStream byteOutputStream = setCustomOutStream();
		
		Game game = new Game();
		addTwoPlayers(game);
		// first turn
		game.roll(11);
		game.wrongAnswer();
		game.roll(1);
		game.wrongAnswer();
		// second turn
		game.roll(1);
		game.wasCorrectlyAnswered();
		game.roll(2);
		game.wasCorrectlyAnswered();
		
		String output = getOutput(byteOutputStream);
		
		Approvals.verify(output);
	}
	
	@Test
	public void testTwoTurnsWithOneWrongAnswerForFirstPlayerAndNotLeavingPenaltyBoxAndRollEqualEleven() {
		ByteArrayOutputStream byteOutputStream = setCustomOutStream();
		
		Game game = new Game();
		addTwoPlayers(game);
		// first turn
		game.roll(11);
		game.wrongAnswer();
		game.roll(1);
		game.wasCorrectlyAnswered();
		// second turn
		game.roll(2);
		game.wasCorrectlyAnswered();
		game.roll(2);
		game.wasCorrectlyAnswered();
		
		String output = getOutput(byteOutputStream);
		
		Approvals.verify(output);
	}
	
	@Test
	public void shouldPlayerLoseWhenPurseEqualsSix() {
		ByteArrayOutputStream byteOutputStream = setCustomOutStream();
		
		Game game = new Game();
		addTwoPlayers(game);

		for(int i = 0; i < 20; i++) { // roll 12 times
			game.roll(1);
			game.wasCorrectlyAnswered();
		}
		
		String output = getOutput(byteOutputStream);
		
		Approvals.verify(output);
	}

	private String getOutput(ByteArrayOutputStream byteOutputStream) {
		System.out.flush();
		String output = byteOutputStream.toString();
		return output;
	}

	private ByteArrayOutputStream setCustomOutStream() {
		ByteArrayOutputStream byteOutputStream = new ByteArrayOutputStream();
		PrintStream printStream = new PrintStream(byteOutputStream);
		System.setOut(printStream);
		return byteOutputStream;
	}
	
	private void addTwoPlayers(Game game) {
		game.add("Player 1");
		game.add("Player 2");
	}
}
