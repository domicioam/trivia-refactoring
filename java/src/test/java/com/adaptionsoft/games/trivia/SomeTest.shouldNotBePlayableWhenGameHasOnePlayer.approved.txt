Player 1 was added
They are player number 1
